lib/travis/api/build/app.rb

```diff
@@ -37,6 +27,35 @@ class App < Sinatra::Base
           if ENV.key?('LIBRATO_EMAIL') && ENV.key?('LIBRATO_TOKEN') && ENV.key?('LIBRATO_SOURCE')
             use Metriks
           end
+
+          use Rack::Deflater
+        end
+
+        helpers do
+          def auth_disabled?
+            (
+              ENV['API_TOKEN'].nil? || ENV['API_TOKEN'].strip.empty?
+            ) && (
+              settings.development? || settings.testing?
+            )
+          end
+        end
+
+        before '/script' do
+          return if auth_disabled?
+
+          unless env.key?('HTTP_AUTHORIZATION')
+            halt 401, 'missing Authorization header'
+          end
+
+          type, token = env['HTTP_AUTHORIZATION'].to_s.split(' ', 2)
+
+          ENV['API_TOKEN'].split(',').each do |valid_token|
+            return if Rack::Utils.secure_compare(type, 'token') &&
+                      Rack::Utils.secure_compare(token, valid_token)
```

Comment:
From the docs for `secure_compare`:

> NOTE: the values compared should be of fixed length, such as strings that have already been processed by HMAC. This should not be used on variable length plaintext strings because it could leak length info via timing attacks.

I'm not sure if leaking length info is a bad thing, though? And if it's something we can avoid?